### ISSUE TITLE
Typechecking performance regression in 23.7.12 with typescript v4 types

### ISSUE THREAD

**Issue by Greg-NetDuma** [i18next/i18next#2100]
## 🐛 Bug Report

Simple, it went from being 65-68s to 160s in our project (after I increased node memory limit to 8G) .

## To Reproduce

I'll get a reproduce repo or something similar soon, but considering we are using Typescript 5 and only the v4 types were changed (https://github.com/i18next/i18next/commit/f9e5971d54379e7529e95b384997ff82884da167) in this release I'm concerned our project is misconfigured. 

## Expected behavior

Performance should be the same as in 23.7.11.

## Your Environment

- *runtime version*: node v21.2.0, typescript 5.2.2
- *i18next version*: 23.7.12
- *os*: Linux

**Comment by Greg-NetDuma** [comment]
It seems the v4 types are used in our project instead of v5, it this line correct? https://github.com/i18next/i18next/blob/70bfcd9a417edce1e0b5ca9191cb7c46518e423e/package.json#L7

Especially that you have the explicit check just below:

```
"typesVersions": {
    "<5.0": {
      "typescript/t.d.ts": [
        "typescript/t.v4.d.ts"
      ],
      "index.d.ts": [
        "index.v4.d.ts"
      ]
    }
  },
```

**Comment by Greg-NetDuma** [comment]
If it's only in v4 types I don't actually care that much but the regression is still there.

**Comment by esetnik** [comment]
I'm not sure this is related but I'm getting a JavaScript heap out of memory using typescript-eslint when using `i18next@23.7.12` or `i18next@23.7.13`. Working if I downgrade to `i18next@23.7.11`. This is using typescript v5. 

$ eslint . --ext ts,tsx,js,jsx,mjs --max-warnings 0

<--- Last few GCs --->

[667:0x6b817d0]   119086 ms: Scavenge (reduce) 2036.0 (2077.4) -> 2035.2 (2076.6) MB, 16.00 / 0.00 ms  (average mu = 0.341, current mu = 0.352) allocation failure; 
[667:0x6b817d0]   119155 ms: Scavenge (reduce) 2035.9 (2076.6) -> 2035.4 (2077.1) MB, 13.91 / 0.00 ms  (average mu = 0.341, current mu = 0.352) allocation failure; 
[667:0x6b817d0]   119251 ms: Scavenge (reduce) 2036.3 (2077.1) -> 2035.8 (2077.6) MB, 6.85 / 0.00 ms  (average mu = 0.341, current mu = 0.352) allocation failure; 


<--- JS stacktrace --->

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0xc9e850 node::Abort() [/usr/local/bin/node]
 2: 0xb720ff  [/usr/local/bin/node]
 3: 0xec1a70 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [/usr/local/bin/node]
 4: 0xec1d57 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [/usr/local/bin/node]
 5: 0x10d3dc5  [/usr/local/bin/node]
 6: 0x10d4354 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/usr/local/bin/node]
 7: 0x10eb244 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*) [/usr/local/bin/node]
 8: 0x10eba5c v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
 9: 0x10edbba v8::internal::Heap::HandleGCRequest() [/usr/local/bin/node]
10: 0x1058fd7 v8::internal::StackGuard::HandleInterrupts() [/usr/local/bin/node]
11: 0x14fa98a v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
12: 0x7f4c6fed9ef6 
Aborted (core dumped)
error Command failed with exit code 134.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

Exited with code exit status 134

**Comment by Greg-NetDuma** [comment]
> I'm not sure this is related but I'm getting a JavaScript heap out of memory using typescript-eslint when using `i18next@23.7.12` or `i18next@23.7.13`. Working if I downgrade to `i18next@23.7.11`. This is using typescript v5.
> 
> $ eslint . --ext ts,tsx,js,jsx,mjs --max-warnings 0
> 
> <--- Last few GCs --->
> 
> [667:0x6b817d0] 119086 ms: Scavenge (reduce) 2036.0 (2077.4) -> 2035.2 (2076.6) MB, 16.00 / 0.00 ms (average mu = 0.341, current mu = 0.352) allocation failure; [667:0x6b817d0] 119155 ms: Scavenge (reduce) 2035.9 (2076.6) -> 2035.4 (2077.1) MB, 13.91 / 0.00 ms (average mu = 0.341, current mu = 0.352) allocation failure; [667:0x6b817d0] 119251 ms: Scavenge (reduce) 2036.3 (2077.1) -> 2035.8 (2077.6) MB, 6.85 / 0.00 ms (average mu = 0.341, current mu = 0.352) allocation failure;
> 
> <--- JS stacktrace --->
> 
> FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory 1: 0xc9e850 node::Abort() [/usr/local/bin/node] 2: 0xb720ff [/usr/local/bin/node] 3: 0xec1a70 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [/usr/local/bin/node] 4: 0xec1d57 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, v8::OOMDetails const&) [/usr/local/bin/node] 5: 0x10d3dc5 [/usr/local/bin/node] 6: 0x10d4354 v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/usr/local/bin/node] 7: 0x10eb244 v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::internal::GarbageCollectionReason, char const*) [/usr/local/bin/node] 8: 0x10eba5c v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node] 9: 0x10edbba v8::internal::Heap::HandleGCRequest() [/usr/local/bin/node] 10: 0x1058fd7 v8::internal::StackGuard::HandleInterrupts() [/usr/local/bin/node] 11: 0x14fa98a v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node] 12: 0x7f4c6fed9ef6 Aborted (core dumped) error Command failed with exit code 134. info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
> 
> Exited with code exit status 134

Try with  NODE_OPTIONS="--max_old_space_size=8192" (or 4096)

Also by the looks of it currently everyone is using the Typescript V4 types .

If you use this patch
```diff
diff --git a/node_modules/i18next/package.json b/node_modules/i18next/package.json
index 830b030..3ff6197 100644
--- a/node_modules/i18next/package.json
+++ b/node_modules/i18next/package.json
@@ -4,7 +4,7 @@
   "description": "i18next internationalization framework",
   "main": "./dist/cjs/i18next.js",
   "module": "./dist/esm/i18next.js",
-  "types": "./index.v4.d.ts",
+  "types": "./index.d.ts",
   "typesVersions": {
     "<5.0": {
       "typescript/t.d.ts": [
```

in theory you should have the same perf as before.

**Comment by esetnik** [comment]
Thanks for the suggestions, but I'd prefer to wait for a proper solution rather than having to apply a patch on `node_modules` as we are using a docker environment.

**Comment by gregorybolkenstijn** [comment]
`23.7.15` has the same out-of-memory issue.

**Comment by esetnik** [comment]
Fixed in `v23.7.16`. Thanks!

**Comment by gregorybolkenstijn** [comment]
I can confirm `23.7.16` fixes the issue, thanks!
